lightningd/peer_control.c

```diff
@@ -1339,6 +1343,101 @@ static const struct json_command getinfo_command = {
 };
 AUTODATA(json_command, &getinfo_command);
 
+
+static struct command_result *json_setchannelfee(struct command *cmd,
+					 const char *buffer,
+					 const jsmntok_t *obj UNNEEDED,
+					 const jsmntok_t *params)
+{
+	struct json_stream *response;
+	const jsmntok_t *idtok;
+	struct peer *peer;
+	struct channel *channel;
+	struct channel_id cid;
+	struct command_result *res;
+	struct amount_msat *base, base_default;
+	u32 *ppm, ppm_default;
+
+	/* Use daemon config values base/ppm as default */
+	amount_msat_from_u32(&base_default, cmd->ld->config.fee_base);
+	ppm_default = cmd->ld->config.fee_per_satoshi;
+
+	/* Parse the JSON command */
+	if (!param(cmd, buffer, params,
+			p_req("id", param_tok, &idtok),
+			p_opt_def("base", param_msat, &base, base_default),
+			p_opt_def("ppm", param_number, &ppm, ppm_default),
+			NULL))
+		return command_param_failed();
+
+	/* Check base for u32 overflow */
+	if (amount_msat_overflow_u32(*base))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				"Value 'base' exceeds u32 max");
+
+	/* Find the channel */
+	peer = peer_from_json(cmd->ld, buffer, idtok);
+	if (peer) {
+		channel = peer_active_channel(peer);
+		if (!channel)
+			return command_fail(cmd, LIGHTNINGD,
+					"Could not find active channel of peer with that id");
+	} else {
+		res = command_find_channel(cmd, buffer, idtok, &channel, &peer);
+		if (res)
+			return res;
+		/* check channel is found and in valid state */
+		if (!channel)
+			return command_fail(cmd, LIGHTNINGD,
+					"Could not find channel with that id");
+	}
+
+	if (channel->state != CHANNELD_NORMAL)
+		return command_fail(cmd, LIGHTNINGD,
+				"Channel is not in state CHANNELD_NORMAL");
+
+	/* set, notify and save new values */
+	channel->feerate_base = amount_msat_to_u32(*base);
+	channel->feerate_ppm = *ppm;
+
+	/* tell channeld to make a send_channel_update */
+	if (channel->owner && streq(channel->owner->name, "lightning_channeld"))
```

Comment:
out of curiosity, in what cases will the channel->owner not be "lightning_channeld"? i.e. is there a mitigation or message value we should return if this check fails?

Reply:
@niftynei  This is a security measure because I don't know the insights of clightning in detail enough. It may be the case that if state := `CHANNELD_NORMAL` it is always owned by channeld. Now that you ask, this is maybe not necessary in this scenario, or we may want a check that enforces it just like the `CHANNELD_NORMAL` a few lines earlier.

I had a ton of open detail questions that were never addressed, one of them being the question if we should allow setting feerates no channel in states other than `CHANNELD_NORMAL`.  In this case a statement like this makes sense, because we only wire the new fees if the channel is online.

So maybe my question to you: In which channel states are we supposed to allow the feerate settings? Because the DB update can maybe be made earlier and the channel will apply correct fees when it is in state `CHANNELD_NORMAL` eventually.

Reply:
well. let's see. the possible channel states are (from `lightningd/channel_state.h):

```lightningd/channel_state.h
       /* In channeld, still waiting for lockin. */
        CHANNELD_AWAITING_LOCKIN = 2,

        /* Normal operating state. */
        CHANNELD_NORMAL,

        /* We are closing, pending HTLC resolution. */
        CHANNELD_SHUTTING_DOWN,

        /* Exchanging signatures on closing tx. */
        CLOSINGD_SIGEXCHANGE,

        /* Waiting for onchain event. */
        CLOSINGD_COMPLETE,

        /* Waiting for unilateral close to hit blockchain. */
        AWAITING_UNILATERAL,

        /* We've seen the funding spent, we're waiting for onchaind. */
        FUNDING_SPEND_SEEN,

        /* On chain */
        ONCHAIN
```

of these, `CHANNELD_NORMAL` and `CHANNELD_AWAITING_LOCKIN` appear to be reasonable candidates for permitting updates to the configured channel. in theory* there's no reason that you _couldn't_ update the channel fees in any of the other states, except that they would have zero effect since the channel is no longer accepting new htlc's (ie those fee updates will never be used).

re: the restriction to only channels that are owned by `channeld`, imo this is coding too tightly to the implementation. the simple heuristic of using well defined channel states to restrict the window of updates is much more flexible, imo. in other words, if we've checked the channel is in the correct state, we should assume that the update is valid.

*ignoring some implementation reality that i'm unaware of that makes this infeasible

Reply:
Thanks for some input ;)
Will include waiting channels without wiring the fees to an otherwise non existent channeld. Write test to see if it works as expected and correct fees will be applied when the channel is in state NORMAL eventually.